kafka-instrumentation/src/main/java/dev/responsive/kafka/clients/EndOffsetsPoller.java

```diff
@@ -0,0 +1,220 @@
+    LOGGER.info("add end offset metrics for thread {}", threadId);
+    final var tm = new ThreadMetrics(threadId);
+    synchronized (this) {
+      assert !threadIdToMetrics.containsKey(threadId);
+      if (threadIdToMetrics.isEmpty()) {
+        initPoller();
+        threadIdToMetrics.put(threadId, tm);
+      }
+    }
+    return tm;
+  }
+
+  public void removeForThread(final String threadId) {
+    final ThreadMetrics tm;
+    synchronized (this) {
+      tm = threadIdToMetrics.remove(threadId);
+      if (tm != null) {
+        tm.cleanup();
+        if (threadIdToMetrics.isEmpty()) {
+          stopPoller();
+        }
+      }
+    }
+  }
+
+  private void pollEndOffsets() {
+    final Set<TopicPartition> partitions = new HashSet<>();
+    final Collection<ThreadMetrics> threadMetrics;
+    synchronized (this) {
+      threadMetrics = List.copyOf(this.threadIdToMetrics.values());
+    }
+    for (final var tm : threadMetrics) {
+      partitions.addAll(tm.endOffsets.keySet());
+    }
+    final var endOffsets = consumer.endOffsets(partitions);
+    threadMetrics.forEach(tm -> tm.update(endOffsets));
+  }
+
+  private void initPoller() {
+    LOGGER.info("init end offsets poller");
+    assert consumer == null;
+    assert executor == null;
+    consumer = factories.createConsumer(configs);
+    executor = factories.createExecutor();
+    pollFuture = executor.scheduleAtFixedRate(this::pollEndOffsets, 0, 30, TimeUnit.SECONDS);
+  }
+
+  private void stopPoller() {
+    LOGGER.info("stopping end offsets poller");
+    try {
+      consumer.close(Duration.ofNanos(0));
+    } catch (final Exception e) {
+      LOGGER.warn("error closing consumer", e);
+    }
+    consumer = null;
+    pollFuture.cancel(true);
+    executor.shutdown();
+    executor = null;
+  }
+
+  private class ThreadMetrics implements ResponsiveConsumer.Listener {
+    private final String threadId;
+    private final Map<TopicPartition, Long> endOffsets = new ConcurrentHashMap<>();
+
+    private ThreadMetrics(final String threadId) {
+      this.threadId = Objects.requireNonNull(threadId);
+    }
+
+    @Override
+    public void onPartitionsLost(final Collection<TopicPartition> lost) {
+      onPartitionsRevoked(lost);
+    }
+
+    @Override
+    public void onPartitionsRevoked(final Collection<TopicPartition> revoked) {
+      for (final var p : revoked) {
+        metrics.removeMetric(metricName(p));
+        endOffsets.remove(p);
+      }
+    }
+
+    @Override
+    public void onPartitionsAssigned(final Collection<TopicPartition> assigned) {
+      for (final var p : assigned) {
+        endOffsets.put(p, -1L);
+        metrics.addMetric(
+            metricName(p),
+            (Gauge<Long>) (config, now) -> endOffsets.getOrDefault(p, -1L)
+        );
+      }
+    }
+
+    private void update(final Map<TopicPartition, Long> endOffsets) {
+      for (final var tp : endOffsets.keySet()) {
+        this.endOffsets.computeIfPresent(tp, (k, v) -> {
```

Comment:
why are we only computing if present here? wouldn't we want to add all endOffsets to the map?

Reply:
End offsets has all the end offsets that were polled (for all threads).